src/badguy/zeekling.cpp

```diff
@@ -54,149 +69,199 @@ Zeekling::collision_squished(GameObject& object)
 
-    // Guess where the player would be at this time.
-    float estPx = (player_pos.x + (estFrames * player_mov.x));
+  auto* resultobj = std::get_if<CollisionObject*>(&result.hit);
+  
+  if (result.is_valid && resultobj &&
+      *resultobj == player->get_collision_object())
+  {
+    m_target_y = player->get_bbox().get_top() + CATCH_OFFSET;
+    return true;
+  }
 
-    // Guess where we would be at this time.
-    float estBx = (self_pos.x + (estFrames * self_mov.x));
+  return false;
+}
 
-    // Allow for slight inaccuracies.
-    if (fabsf(estPx - estBx) < 8) return true;
-  }
+void
+Zeekling::set_speed(float speed)
+{
+  m_physic.set_velocity_x(speed * (m_dir == Direction::LEFT ? -1 : 1));
+}
 
-  // Update the last player tracked, as well as our positions.
-  last_player = player;
-  if (player) {
-    last_player_pos = player->get_pos();
-    last_self_pos = m_col.m_bbox.p1();
-  }
+void Zeekling::fly()
+{
+  m_state = FLYING;
+  set_speed(FLYING_SPEED);
+  m_start_position.y = get_pos().y;
+  set_action(m_dir);
+}
 
-  return false;
+void Zeekling::charge()
+{
+  m_state = CHARGING;
+  m_timer.start(CHARGING_DURATION);
+  set_speed(CHARGING_SPEED);
+  set_action("charge", m_dir);
+}
+
+void Zeekling::dive()
+{
+  m_state = DIVING;
+  m_timer.start(DIVING_DURATION);
+  set_speed(DIVING_SPEED);
+  set_action("dive", m_dir);
+}
+
+void Zeekling::recover()
+{
+  m_state = RECOVERING;
+  m_catch_pos = get_pos().y;
+  m_timer.start(RECOVER_DURATION);
+  set_action(m_dir);
 }
 
 void
 Zeekling::active_update(float dt_sec) {
-  if (state == FLYING) {
-    if (should_we_dive()) {
-      state = DIVING;
-      m_physic.set_velocity_y(2*fabsf(m_physic.get_velocity_x()));
-      set_action("dive", m_dir);
-    }
-    BadGuy::active_update(dt_sec);
-    return;
-  } else if (state == DIVING) {
-    BadGuy::active_update(dt_sec);
-    return;
-  } else if (state == CLIMBING) {
-    // Stop climbing when we're back at initial height.
-    if (get_pos().y <= m_start_position.y) {
-      state = FLYING;
-      m_physic.set_velocity_y(0);
-    }
-    BadGuy::active_update(dt_sec);
-    return;
-  } else {
-    assert(false);
+  switch (m_state) {
+    case FLYING:
+      if (!should_we_dive())
+        break;
+
+      charge();
+
+      break;
+
+    case CHARGING:
+      if (m_timer.check())
+      {
+        dive();
+      }
+      break;
+
+    case DIVING:
+      if (m_timer.check())
+      {
+        recover();
+      }
+      else
+      {
+        const float dist = m_target_y - m_start_position.y;
+        const double progress = CubicEaseIn(static_cast<double>(1.f - m_timer.get_progress()));
+        const float value = m_target_y - (static_cast<float>(progress) * dist);
+        const Vector pos(get_pos().x, value);
+        set_pos(pos);
+      }
+      break;
+
+    case RECOVERING:
+      if (m_timer.check())
+      {
+        fly();
+      }
+      else
+      {
```

Comment:
Maybe you can get rid of the else block by `break;`-ing early in the RECOVERING case.

Reply:
thanks for pointing that out